// defender-relay-client/lib/autotask/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AutotaskRelayer = void 0;
const autotask_1 = require("defender-base-client/lib/autotask");
class AutotaskRelayer extends autotask_1.BaseAutotaskClient {
    constructor(params) {
        super(params.credentials, params.relayerARN);
        this.jsonRpcRequestNextId = 0;
    }
    async sendTransaction(payload) {
        return this.execute({ action: 'send-tx', payload });
    }
    async replaceTransactionById(transactionId, txPayload) {
        const payload = { ...txPayload, transactionId };
        return this.execute({ action: 'replace-tx', payload });
    }
    async replaceTransactionByNonce(nonce, txPayload) {
        const payload = { ...txPayload, nonce };
        return this.execute({ action: 'replace-tx', payload });
    }
    async getRelayer() {
        return this.execute({
            action: 'get-self',
        });
    }
    async query(id) {
        return this.execute({
            action: 'get-tx',
            payload: id,
        });
    }
    async list(criteria) {
        return this.execute({
            action: 'list-txs',
            payload: criteria !== null && criteria !== void 0 ? criteria : {},
        });
    }
    async sign(payload) {
        return this.execute({
            action: 'sign',
            payload: payload,
        });
    }
    async call(method, params) {
        return this.execute({
            action: 'json-rpc-query',
            payload: { method, params, jsonrpc: '2.0', id: this.jsonRpcRequestNextId++ },
        });
    }
}
exports.AutotaskRelayer = AutotaskRelayer;
