// defender-autotask-client/lib/api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AutotaskClient = void 0;
const defender_base_client_1 = require("defender-base-client");
const zip_1 = require("./zip");
class AutotaskClient extends defender_base_client_1.BaseApiClient {
    getPoolId() {
        return process.env.DEFENDER_AUTOTASK_POOL_ID || 'us-west-2_94f3puJWv';
    }
    getPoolClientId() {
        return process.env.DEFENDER_AUTOTASK_POOL_CLIENT_ID || '40e58hbc7pktmnp9i26hh5nsav';
    }
    getApiUrl() {
        return process.env.DEFENDER_AUTOTASK_API_URL || 'https://defender-api.openzeppelin.com/autotask/';
    }
    async updateCodeFromZip(autotaskId, zippedCode) {
        const encodedZippedCode = zippedCode.toString('base64');
        return this.updateCode(autotaskId, encodedZippedCode);
    }
    async updateCodeFromSources(autotaskId, sources) {
        const encodedZippedCode = await zip_1.zipSources(sources);
        return this.updateCode(autotaskId, encodedZippedCode);
    }
    async updateCodeFromFolder(autotaskId, path) {
        const encodedZippedCode = await zip_1.zipFolder(path);
        return this.updateCode(autotaskId, encodedZippedCode);
    }
    async updateCode(autotaskId, encodedZippedCode) {
        return this.apiCall(async (api) => {
            return await api.put(`/autotasks/${autotaskId}/code`, { encodedZippedCode });
        });
    }
}
exports.AutotaskClient = AutotaskClient;
